url_shortener/views/handlers.py

```diff
@@ -1,60 +1,332 @@
+
+    logic: Logic = request.registry.logic
+
+    if logic.remove_token(get_auth_token(request)):
+
+        return Response(status=httplib.OK, json_body={
+            'status': 'Success',
+            'description': 'Successfully logged out.'
+        })
+    else:
+        return Response(status=httplib.CONFLICT, json_body={
+            'status': 'error',
+            'description': 'Smth went wrong.'
+        })
 
-def protected_resource_write_example(request: Request) -> Response:
-    key = request.matchdict['key']
-    value = request.json_body.get('value')
 
-    if value is None:
-        return Response(status=httplib.BAD_REQUEST,
-                        json_body={
-                            'status': 'error',
-                            'reason':
-                            '`value` was not provided within request',
-                        })
+def redirect_longlink(request: Request):
+    """ Redirects if conditions are met.
+        CONDITIONS:
+            - If data entered is non-null
+            - If key exists.
+            - If key is valid.
+        Returns HTTPFound
+    """
 
     logic: Logic = request.registry.logic
-    success = logic.save_example_if_not_exists(key, value)
+    try:
+        id = request.matchdict['id']
+    except Exception as e:
+        print(e)
+        return Response(status=httplib.BAD_REQUEST, json_body={
+            'status': 'error',
+            'description': 'Request body is missing.'
+        })
+
+    try:
+        long_link = logic.read_by_key(id)
+    except Exception as e:
+        print(e)
+        return Response(status=httplib.BAD_REQUEST, json_body={
+            'status': 'error',
+            'description': 'Key does not exist.'
+        })
+
+    if id is None or long_link is None:
+        return Response(status=httplib.BAD_REQUEST, json_body={
+            'status': 'error',
+            'description': 'Incorrect key.'
+        })
 
+    return HTTPFound(location=long_link)
+
+
+def create_shortlink(request: Request) -> Response:
+    """ Creates a shortlink id, by using id if provided or generating 4 char random string."""
+    logic: Logic = request.registry.logic
+
+    try:
+        url_provided = request.json_body.get('url')
+        id_provided = request.json_body.get('id')
+    except Exception as e:
+        print(e)
+        return Response(status=httplib.BAD_REQUEST, json_body={
+            'status': 'error',
+            'description': 'Request body is missing.'
+        })
+
+    if url_provided is None:
+        return Response(status=httplib.BAD_REQUEST, json_body={
+            'status': 'error',
+            'description': 'URL must be provided.'
+        })
+
+    # If custom id is not provided, or this id already exists, we give an random one.
+    while id_provided is None or logic.read_by_key(id_provided) is not None:
+        id_provided = ''.join(random.choice(string.ascii_uppercase + string.ascii_lowercase) for _ in range(4))
```

Comment:
1. This looks like a part of business logic. Should it really be in the handler, or should it be in the `logic` module?
2. When you use all 4-character strings your application will stuck here in the infinite loop :(